### ISSUE TITLE
Offscrene rendering to preallocated buffer

### ISSUE THREAD

**Issue by hmaarrfk** [vispy/vispy#1959]
When rendering larger visuals, it might be useful to be able to render the images to a pre-allocated buffer. Would this be an acceptable feature to support in vispy?

If so, I would be happy to help get this into vispy.

**Comment by djhoese** [comment]
I wouldn't mind adding support for it, but it depends exactly what you're thinking. At some level the double buffer rending that OpenGL uses by default (or at least the way vispy configures things) should mean that the next "frame" is already being rendered while viewing the current one. The buffers are just swapped then when the next frame is to be shown.

There are some visuals like the TextVisual which pre-render characters into a TextureAtlas (there is a CPU renderer or a GPU renderer) using this concept of "render to a framebuffer and then put it somewhere".

**Comment by hmaarrfk** [comment]
Specifically, i want to pass my own buffer to:

https://github.com/vispy/vispy/blob/7f3f6f60c8462bb8a3a8fa03344a2e6990b86eb2/vispy/gloo/wrappers.py#L604

Currently, it is using `ctypes.createstringbuffer`

but I would like to recycle my own buffer
https://github.com/vispy/vispy/blob/7f3f6f60c8462bb8a3a8fa03344a2e6990b86eb2/vispy/gloo/gl/_gl2.py#L979

**Comment by hmaarrfk** [comment]
The es version uses the same idea:
https://github.com/vispy/vispy/blob/7f3f6f60c8462bb8a3a8fa03344a2e6990b86eb2/vispy/gloo/gl/_es2.py#L733

**Comment by djhoese** [comment]
I suppose this could be seen as something like an in-place read_pixels option? It'd be nice if you could provide a write-able numpy array that was handled correctly and filled in by the glReadPixels. Putting this another way, you're saying you don't want glReadPixels to make a new output/CPU-side buffer every time you read the pixels, right?

**Comment by hmaarrfk** [comment]
You are correct, I would like to provide my own numpy buffer for rendering avoiding the allocation of a potentially large CPU buffer.

**Comment by djhoese** [comment]
Sounds good to me. I await your PR :wink:

**Comment by djhoese** [comment]
I added the `performance` label to this because it overall would effect performance if used correctly and it made me wonder, do you think there are any parts of vispy that might benefit from this?

**Comment by hmaarrfk** [comment]
i'm not too sure if there are other places that might benefit. And to be honest, we aren't sure if this will help performance, but we are happy to report with simple benchmarks.

This kind of stuff doesn't help too much with 10MB arrays (which are of concern for 1080 screens) but become important for larger datasizes, (which we are very interested in).

**Comment by djhoese** [comment]
So like 4K displays? Because read_pixels should only care about the size of the canvas/buffer, right? If you can tell me more about what you're working on, I'm curious.

**Comment by hmaarrfk** [comment]
We are making gigapixel microscopy tools, from hardware to software.

Part of it involves visualization, so sometimes our customers do wish to see the data they are acquiring. Go figure!

We use python quite extensively to prototype and vispy has allowed us to visualize things quite nicely. We are looking to export certain screenshots more programmatically and looking into the possibility of using vispy for that.

We actually just got a 4k display, but vispy works for that natively.
What we are interested in this discussion is more rendering to RAM, or eventually, to disk.

I couldn't find your email easily, but i'm happy to discuss this in a more private forum. I think you might find it interesting.
